nnfabrik/templates/checkpoint.py

```diff
@@ -0,0 +1,166 @@
+import copy
+import os
+import tempfile
+
+import datajoint as dj
+import torch
+
+from nnfabrik.templates.trained_model import TrainedModelBase
+from nnfabrik.utility.dj_helpers import make_hash, clone_conn, CustomSchema
+
+
+def my_checkpoint(nnfabrik):
+    conn_clone = clone_conn(dj.conn())
+    schema_clone = CustomSchema(nnfabrik.schema.database, connection=conn_clone)
+
+    @schema_clone
+    class Checkpoint(dj.Manual):
+        storage = "minio"
+
+        @property
+        def definition(self):
+            definition = f"""
+            # Checkpoint table
+            -> nnfabrik.Trainer
+            -> nnfabrik.Dataset
+            -> nnfabrik.Model
+            -> nnfabrik.Seed
+            epoch:                             int          # epoch of creation
+            ---
+            score:                             float        # current score at epoch
+            state:                             attach@{self.storage}  # current state
+            ->[nullable] nnfabrik.Fabrikant
+            trainedmodel_ts=CURRENT_TIMESTAMP: timestamp    # UTZ timestamp at time of insertion
+            """
+            return definition
+
+    return Checkpoint
+
+
+class TrainedModelChkptBase(TrainedModelBase):
+    checkpoint_table = None  # Checkpoint
+    keys = [
+        "model_fn",
+        "model_hash",
+        "dataset_fn",
+        "dataset_hash",
+        "trainer_fn",
+        "trainer_hash",
+    ]
+
+    def call_back(self, uid=None, epoch=None, model=None, state=None):
+        """
+        This method is periodically called by the trainer and is used to save the training state in a remote table.
+        Args:
+            uid - Unique identifier for the trained model entry
+            epoch - the iteration count
+            model - current model under training
+            state - Additional information provided by the trainer
+                action: str = "save"/"last"/"best"
+                score: float = 0.0,
+                maximize_score: bool = True,
+                save_every_n: int = 1,
+                keep_last_n: int = 1,
+                keep_best_n: int = 1,
+                keep_selection: Tuple = (),
+        """
+        maximize_score = state.pop("maximize_score", True)
+        action = state.pop("action", "save")
+        if action == "save":  # save current epoch
+            self.save_epoch(epoch, maximize_score, model, state, uid)
+        else:  # restore last/best existing epoch
+            self.restore_epoch(action, uid, maximize_score, state)
+
+    def restore_epoch(self, action, uid, maximize_score, state):
+        # retrieve all fitting entries from checkpoint table
```

Comment:
I am not a big fan of comments that just explain what the code following the comment does. Generally the necessity for such comments indicates that the code could be written in a way that is easier to understand and would therefore eliminate the need for the comment. In this case I think the comment can be removed because the code clearly expresses what it does.